clients/pkg/promtail/targets/gcplog/push_target_test.go

```diff
@@ -0,0 +1,304 @@
+		"messageId": "5187581549398349",
+		"message_id": "5187581549398349",
+		"publishTime": "2022-07-25T22:19:15.56Z",
+		"publish_time": "2022-07-25T22:19:15.56Z"
+	},
+	"subscription": "projects/wired-height-350515/subscriptions/test"
+}`
+
+func makeGCPPushRequest(host string, body string) (*http.Request, error) {
+	req, err := http.NewRequest(http.MethodPost, fmt.Sprintf("%s/gcp/api/v1/push", host), strings.NewReader(body))
+	if err != nil {
+		return nil, err
+	}
+	return req, nil
+}
+
+func TestPushTarget(t *testing.T) {
+	w := log.NewSyncWriter(os.Stderr)
+	logger := log.NewLogfmtLogger(w)
+
+	type expectedEntry struct {
+		labels model.LabelSet
+		line   string
+	}
+	type args struct {
+		RequestBody    string
+		RelabelConfigs []*relabel.Config
+		Labels         model.LabelSet
+	}
+
+	cases := map[string]struct {
+		args            args
+		expectedEntries []expectedEntry
+	}{
+		"simplified cloud functions log line": {
+			args: args{
+				RequestBody: testPayload,
+				Labels: model.LabelSet{
+					"job": "some_job_name",
+				},
+			},
+			expectedEntries: []expectedEntry{
+				{
+					labels: model.LabelSet{
+						"job": "some_job_name",
+					},
+					line: expectedMessageData,
+				},
+			},
+		},
+		"simplified cloud functions log line, with relabeling custom attribute and message id": {
+			args: args{
+				RequestBody: testPayload,
+				Labels: model.LabelSet{
+					"job": "some_job_name",
+				},
+				RelabelConfigs: []*relabel.Config{
+					{
+						SourceLabels: model.LabelNames{"__gcp_attributes_logging_googleapis_com_timestamp"},
+						Regex:        relabel.MustNewRegexp("(.*)"),
+						Replacement:  "$1",
+						TargetLabel:  "google_timestamp",
+						Action:       relabel.Replace,
+					},
+					{
+						SourceLabels: model.LabelNames{"__gcp_message_id"},
+						Regex:        relabel.MustNewRegexp("(.*)"),
+						Replacement:  "$1",
+						TargetLabel:  "message_id",
+						Action:       relabel.Replace,
+					},
+				},
+			},
+			expectedEntries: []expectedEntry{
+				{
+					labels: model.LabelSet{
+						"job":              "some_job_name",
+						"google_timestamp": "2022-07-25T22:19:09.903683708Z",
+						"message_id":       "5187581549398349",
+					},
+					line: expectedMessageData,
+				},
+			},
+		},
+	}
+	for name, tc := range cases {
+		t.Run(name, func(t *testing.T) {
+			// Create fake promtail client
+			eh := fake.New(func() {})
+			defer eh.Stop()
+
+			serverConfig, port, err := getServerConfigWithAvailablePort()
+			require.NoError(t, err, "error generating server config or finding open port")
+			config := &scrapeconfig.GcplogTargetConfig{
+				Server:               serverConfig,
+				Labels:               tc.args.Labels,
+				UseIncomingTimestamp: false,
+				SubscriptionType:     "push",
+			}
```

Comment:
Running the server is quite a handful. What do you think about testing the handler function `push` with a [ResponseRecorder](https://pkg.go.dev/net/http/httptest#ResponseRecorder) instead?